Refactor ESQL query generation to support selective string casting

simianhacker opened the pull request:
## 🍒 Summary

This pull request refactors the ESQL query generation in the `kbn-unified-metrics-grid` package to support selective string casting for non-keyword dimension fields. This prevents query failures when using IP or numeric dimensions while maintaining optimal performance by not casting keyword fields.

Fixes #236337

## 🛠️ Changes

- Updated the `createESQLQuery` function to accept a full `MetricField` object, providing access to dimension type information.
- Implemented a `needsStringCasting` helper function to identify field types (`ip`, `long`, `integer`, `boolean`, etc.) that require `::STRING` casting in ESQL.
- Modified the `CONCAT` logic to selectively apply casting only to the identified non-keyword fields.
- Updated the component call site in `chart/index.tsx` to pass the entire `metric` object.
- Expanded the Jest test suite to include comprehensive test cases for various casting scenarios, including mixed field types and edge cases.


## 📷 Screenshots

<img width="2048" height="1006" alt="image" src="https://github.com/user-attachments/assets/40a3e6e0-cc39-4dfc-b1a8-e9c68d4c029f" />

<img width="2045" height="1002" alt="image" src="https://github.com/user-attachments/assets/ea651b7b-1112-449d-9f62-3386106fe783" />

## 🎙️ Prompt

I'm looking into this bug in Kibana: https://github.com/elastic/kibana/issues/236337 

The code that creates the ESQL query for the Lens visualizations is in the `kbn-unified-metrics-grid` platform package. When the user selects 2 dimensions (`host.ip` and `host.name`) the metrics grid produces an ESQL like:

```
TS metrics-hostmetricsreceiver.otel-default
  | WHERE host.ip IS NOT NULL AND host.name IS NOT NULL
  | STATS AVG(system.cpu.load_average.`1m`) BY BUCKET(@timestamp, 100, ?_tstart, ?_tend), host.ip, host.name
  | EVAL __DIMENSIONS__ = CONCAT(host.ip, " › ", host.name)
  | DROP host.ip, host.name
```

This query will fail with the following message from Elasticsearch: 

```
[esql] > Unexpected error from Elasticsearch: verification_exception - Found 1 problem
line 4:27: argument of [CONCAT(host.ip, " › ", host.name)] must be [string], found value [host.ip] type [ip]
```

This is because `CONCAT` expects 2 strings and concatenates them together. To fix this we should cast the fields like `CONCAT(host.ip::STRING, " › ", host.name)` for non-keyword fields. **We can take a precise approach by using the dimension type information available on `MetricField.dimensions` to selectively cast only non-keyword fields.**

### Additional context

I recently merged this PR (https://github.com/elastic/kibana/pull/235421) to include `ip` and numeric types as available dimensions for the user to pick. 

### Investigation Tips
- Try starting the investigation with `list_symbols_by_query(kql='filePath: *kbn-unified-metrics-grid*')`"

🤖 This pull request was assisted by Gemini CLI

simianhacker committed:
Refactor ESQL query generation to support selective string casting

- Updated `createESQLQuery` to accept a `MetricField` object, enabling access to dimension types.
- Implemented `needsStringCasting` helper to identify non-keyword fields needing `::STRING` casting.
- Modified ESQL `CONCAT` generation to selectively cast only IP, numeric, and boolean dimension fields.
- Updated the call site in the Chart component to pass the full `metric` object.
- Revised and expanded tests to validate selective casting for various field type combinations.

Prompts:

- "## Problem Statement

I'm looking into this bug in Kibana: https://github.com/elastic/kibana/issues/236337

The code that creates the ESQL query for the Lens visualizations is in the `kbn-unified-metrics-grid` platform package. When the user selects 2 dimensions (`host.ip` and `host.name`) the metrics grid produces an ESQL like:

```
TS metrics-hostmetricsreceiver.otel-default
  | WHERE host.ip IS NOT NULL AND host.name IS NOT NULL
  | STATS AVG(system.cpu.load_average.`1m`) BY BUCKET(@timestamp, 100, ?_tstart, ?_tend), host.ip, host.name
  | EVAL __DIMENSIONS__ = CONCAT(host.ip, " › ", host.name)
  | DROP host.ip, host.name
```

This query will fail with the following message from Elasticsearch:

```
[esql] > Unexpected error from Elasticsearch: verification_exception - Found 1 problem
line 4:27: argument of [CONCAT(host.ip, " › ", host.name)] must be [string], found value [host.ip] type [ip]
```

This is because `CONCAT` expects 2 strings and concatenates them together. To fix this we should cast the fields like `CONCAT(host.ip::STRING, " › ", host.name)` for non-keyword fields. **We can take a precise approach by using the dimension type information available on `MetricField.dimensions` to selectively cast only non-keyword fields.**

### Additional context

I recently merged this PR (https://github.com/elastic/kibana/pull/235421) to include `ip` and numeric types as available dimensions for the user to pick.

### **Updated Solution Approach**
Instead of casting all fields to strings, we can:
1. Pass the full `MetricField` object to `createESQLQuery` instead of just `metric.name` and `metric.instrument`
2. Use `metric.dimensions` array to get type information for each dimension
3. Selectively cast only non-keyword types (ip, numeric, boolean) to strings
4. Leave keyword fields uncast for better performance

### Investigation Tips
- Try starting the investigation with `list_symbols_by_query(kql='filePath: *kbn-unified-metrics-grid*')`"

🤖 This commit was assisted by Gemini CLI

elasticmachine commented:
Pinging @elastic/obs-ux-infra_services-team (Team:obs-ux-infra_services)

simianhacker committed:
Add JSDoc blocks to ESQL query generation functions

- Added JSDoc comments to the `createESQLQuery` and `needsStringCasting` functions to improve code documentation and clarity.

Prompts:

- "Can you add a doc block to createESQLQuery and needStringCasting?"

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
Refactor needsStringCasting to use shared constant

- Refactored the `needsStringCasting` helper to use the `DIMENSION_TYPES` constant from the `@kbn/metrics-experience-plugin`.
- Removed the hardcoded `Set` of dimension types, centralizing the logic.

Prompts:

- "For the createESQLQuery's list of dimension types, can you use the const DIMENSION_TYPES in this file src/platform/plugins/shared/metrics_experience/common/fields/constants.ts you should be able to access it like the types from the metrics-experiance plugin."
- "can you run the test for this file?"

🤖 This commit was assisted by Gemini CLI

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/343481)
* Commit: ea57e577fc3440edfb75e7b8c73ef32e7ca27f9d

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/343481#01997ddf-beb4-4679-8d3c-8d8d22f76a81)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/343481","number":343481,"commit":"ea57e577fc3440edfb75e7b8c73ef32e7ca27f9d"}],"number":343481}
buildkite-pr-comment-->

simianhacker committed:
Fix build error by exporting shared constants

- Exported `DIMENSION_TYPES` and `NUMERIC_TYPES` from the `metrics_experience` plugin's public API.
- Updated the import path in `create_esql_query.ts` to consume the newly exposed constants.
- This resolves a module resolution error that occurred during the build process.

Prompts:

- "I'm getting this error..."
- "Won't we still need the ES_FIELD_TYPES?"
- "I fixed it manually... try the test again"

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

simianhacker committed:
moving the dimension types to the common/constants

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

simianhacker committed:
removing a kbn package from tscfonig

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/344510)
* Commit: 65682197a13d4f99a617a067a63acdb92242747a

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/344510#01998741-a204-4015-b593-6cb83d00a0c2)

### History
* :broken_heart: [Build #344057](https://buildkite.com/elastic/kibana-pull-request/builds/344057) failed 626a427f7ec7a20805782cbfefa9181b3019eda5
* :broken_heart: [Build #343814](https://buildkite.com/elastic/kibana-pull-request/builds/343814) failed 79e8f1982659a76355f4d63cc10cb93be811d877
* :broken_heart: [Build #343766](https://buildkite.com/elastic/kibana-pull-request/builds/343766) failed ee86be7fc6182ac9cd98a83c18a6fb1410a023a5
* :broken_heart: [Build #343481](https://buildkite.com/elastic/kibana-pull-request/builds/343481) failed ea57e577fc3440edfb75e7b8c73ef32e7ca27f9d

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/344510","number":344510,"commit":"65682197a13d4f99a617a067a63acdb92242747a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/344057","number":344057,"commit":"626a427f7ec7a20805782cbfefa9181b3019eda5"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/343814","number":343814,"commit":"79e8f1982659a76355f4d63cc10cb93be811d877"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/343766","number":343766,"commit":"ee86be7fc6182ac9cd98a83c18a6fb1410a023a5"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/343481","number":343481,"commit":"ea57e577fc3440edfb75e7b8c73ef32e7ca27f9d"}],"number":344510}
buildkite-pr-comment-->

simianhacker committed:
Grasping at straws trying to get the build to pass

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

simianhacker committed:
shooting in the dark with these build errors... again

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/344583)
* Commit: 9a63f1f24883e8a9ccf547dac69143401bed8a80

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/344583#01998846-7835-4430-8d94-dc23e97b5c3d)

### History
* :broken_heart: [Build #344563](https://buildkite.com/elastic/kibana-pull-request/builds/344563) failed a56e4bea7681d13212e5b2a84b8e333e42f79bcf
* :broken_heart: [Build #344510](https://buildkite.com/elastic/kibana-pull-request/builds/344510) failed 65682197a13d4f99a617a067a63acdb92242747a
* :broken_heart: [Build #344057](https://buildkite.com/elastic/kibana-pull-request/builds/344057) failed 626a427f7ec7a20805782cbfefa9181b3019eda5
* :broken_heart: [Build #343814](https://buildkite.com/elastic/kibana-pull-request/builds/343814) failed 79e8f1982659a76355f4d63cc10cb93be811d877
* :broken_heart: [Build #343766](https://buildkite.com/elastic/kibana-pull-request/builds/343766) failed ee86be7fc6182ac9cd98a83c18a6fb1410a023a5

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/344583","number":344583,"commit":"9a63f1f24883e8a9ccf547dac69143401bed8a80"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/344563","number":344563,"commit":"a56e4bea7681d13212e5b2a84b8e333e42f79bcf"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/344510","number":344510,"commit":"65682197a13d4f99a617a067a63acdb92242747a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/344057","number":344057,"commit":"626a427f7ec7a20805782cbfefa9181b3019eda5"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/343814","number":343814,"commit":"79e8f1982659a76355f4d63cc10cb93be811d877"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/343766","number":343766,"commit":"ee86be7fc6182ac9cd98a83c18a6fb1410a023a5"}],"number":344583}
buildkite-pr-comment-->

jennypavlova reviewed (commented):
Code LGTM 🚀
~~I couldn't test it as when I switched to the branch, bootstraped, and restarted Kibana, I wasn't able to see the results as I landed on the no results page (I see data returned)~~

<img width="920" height="424" alt="image" src="https://github.com/user-attachments/assets/fd1b4dc0-83b0-4a6a-ae37-87fda0147b1d" />


~~I tried both `FROM` and `TS`, and I have the feature flag in my kibana.yml.
Can you please take a look? I will also try a full reset to make sure it's not a problem of my env + switching branches.~~

I see it after reset 🎉 

<img width="1809" height="743" alt="image" src="https://github.com/user-attachments/assets/605266c5-e31a-4ba1-a484-2f6cb34d0992" />

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

simianhacker committed:
Update Dimension interface and related functions to use ES_FIELD_TYPES

- Updated the `Dimension` interface in `types.ts` to use `ES_FIELD_TYPES` for the `type` property, ensuring stricter type checking.
- Modified the `needsStringCasting` function in `create_esql_query.ts` to accept `ES_FIELD_TYPES` for its `fieldType` parameter.
- Adjusted the `extractDimensions` function in `extract_dimensions.ts` to ensure the dimension's type is one of the allowed `DIMENSION_TYPES` and is cast to `ES_FIELD_TYPES`.
- Updated related test files to align with the new type definitions.

Prompts:

- "Can you use the Github MCP tool to read the code comment/feedback in PR 236385 in elastic/kibana repo?"
- "Create a plan to apply her feedback."
- "Can you look at @/Users/ccowan/Projects/kibana/src/platform/plugins/shared/metrics_experience/server/lib/dimensions/extract_dimensions.ts ? on line 36 I need to ensure that's set to the proper ES_FIELD_TYPE for the Dimension which could be one of the dimension listed in DIMENSION_TYPES from @/Users/ccowan/Projects/kibana/src/platform/plugins/shared/metrics_experience/common/fields/constants.ts"

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
replacing strings with ES_FIELD_TYPES

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

jennypavlova commented on src/platform/packages/shared/kbn-unified-metrics-grid/src/common/utils/esql/create_esql_query.ts:
```diff
@@ -9,25 +9,51 @@
+ */
+function needsStringCasting(fieldType: string): boolean {
+  return (
+    DIMENSION_TYPES.includes(fieldType as ES_FIELD_TYPES) && fieldType !== ES_FIELD_TYPES.KEYWORD
```
Thanks for the changes!

jennypavlova reviewed (approved):
LGTM 💯

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/345862)
* Commit: acc7d890fe3d488c309bcd9eb8c167e61c57d733

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/345862#01999b01-f5ab-4007-b624-fd9f380cbc7a)

### History
* :broken_heart: [Build #345427](https://buildkite.com/elastic/kibana-pull-request/builds/345427) failed 438adf7f8b7c814a1a573c0e5dd367e8de7b2d1e
* :broken_heart: [Build #345404](https://buildkite.com/elastic/kibana-pull-request/builds/345404) failed 657e2fac02d7e4b7786a615c7b920d7203f3b9dc
* :broken_heart: [Build #345271](https://buildkite.com/elastic/kibana-pull-request/builds/345271) failed a3835770e5ede9ce8f1cb391445bb5b331ff6526
* :broken_heart: [Build #345230](https://buildkite.com/elastic/kibana-pull-request/builds/345230) failed 0315fd3a0efba01e48051623fdd9514a04c0431a
* :broken_heart: [Build #344583](https://buildkite.com/elastic/kibana-pull-request/builds/344583) failed 9a63f1f24883e8a9ccf547dac69143401bed8a80

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/345862","number":345862,"commit":"acc7d890fe3d488c309bcd9eb8c167e61c57d733"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/345427","number":345427,"commit":"438adf7f8b7c814a1a573c0e5dd367e8de7b2d1e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/345404","number":345404,"commit":"657e2fac02d7e4b7786a615c7b920d7203f3b9dc"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/345271","number":345271,"commit":"a3835770e5ede9ce8f1cb391445bb5b331ff6526"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/345230","number":345230,"commit":"0315fd3a0efba01e48051623fdd9514a04c0431a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/344583","number":344583,"commit":"9a63f1f24883e8a9ccf547dac69143401bed8a80"}],"number":345862}
buildkite-pr-comment-->

simianhacker committed:
Adding @kbn/field-types to tsconfig.json

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

simianhacker committed:
Removing a usless test, index is never undefined

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

simianhacker commented on src/platform/plugins/shared/metrics_experience/common/dimensions/types.ts:
```diff
@@ -7,7 +7,10 @@
   name: string;
-  type: string;
+  type: ES_FIELD_TYPES;
+  description?: string;
```
Whoops... I messed that up in the merge conflict.

simianhacker committed:
removing description, got missed in a merge conflict

simianhacker committed:
Removing 'description' from dimensions

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/esql-selective-string-casting
